src/shared/sleep-config.c

```diff
@@ -108,77 +111,192 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
+                        c = true;
+        }
+        /* if all batteries are below 5 then c will remain false */
+        if (!c)
+                return 1;
+
+        return 0;
+}
+
+/* get sum of all batteries current charge level */
+int get_total_battery_capacity(void) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *dev;
+        int total_battery_capacity, r;
+
+        r = get_sd_device_battery(&e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get available battery: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                r = read_battery_capacity_percentage(dev);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to get battery capacity: %m");
+                        continue;
+                }
+                total_battery_capacity = total_battery_capacity + r;
+        }
+
+        return total_battery_capacity;
+}
+
+/* Store current capacity of each battery before suspension and timestamp */
+int hashmap_battery_name_with_battery_capacity(Hashmap **ret_current_capacity) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        _cleanup_(hashmap_freep) Hashmap *current_capacity = NULL;
+        sd_device *dev;
+        int battery_capacity, r;
+
+        assert(current_capacity);
+
+        current_capacity = hashmap_new(&string_hash_ops);
+        if (!current_capacity)
+                return log_oom();
+
+        r = get_sd_device_battery(&e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get available battery: %m");
+
+        FOREACH_DEVICE(e, dev) { /* For AC, the TYPE is mains */
+                const char *battery_name;
+
+                r = read_battery_capacity_percentage(dev);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to get battery capacity: %m");
+                        continue;
+                }
+                battery_capacity = r;
+
+                r = sd_device_get_property_value(dev, "POWER_SUPPLY_NAME", &battery_name);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery name: %m");
+                        continue;
+                }
+
+                r = hashmap_put(current_capacity, battery_name, INT_TO_PTR(battery_capacity));
+                if (r < 0) {
+                        log_error_errno(r, "Failed to add to hashmap last battery capacity: %m");
+                        continue;
+                }
+        }
+        *ret_current_capacity = TAKE_PTR(current_capacity);
+
+        return 0;
+}
+
 /* Read file path and return hash of value in that file */
-static int get_battery_identifier(const char *filepath, struct siphash *ret) {
-        _cleanup_free_ char *value = NULL;
+static int get_battery_identifier(sd_device *dev, const char *property, struct siphash *ret) {
```

Comment:
In this case, `struct siphash *` is an input, and updated by the function. Hence, please do not use `ret`. Maybe, `state` is better name.

Reply:
yes. right. it is an input. it is updated and then returned.